[REF] Update production deployment docs (#352)
neurobagel/documentation: docs/user_guide/production_deployment.md

you should customize the settings to fit your specific needs.

To help you with this,
the Neurobagel [`recipes` repository](https://github.com/neurobagel/recipes)
includes dedicated production deployment recipes that should
cover the most common use cases.

!!! info "Start with a fresh recipe for each deployment"

    Make a fresh clone of the 
    [`recipes` repository](https://github.com/neurobagel/recipes) for each
    deployment you want to launch.
    
    For example, hosting multiple Neurobagel [nodes](#node),
    or a single [node](#node) alongside a [containerized proxy server](#proxy-server),
    means creating a separate clone of the `recipes` repo for each node or proxy server.
    This will make it easier to update and maintain each deployment.

## Deployable services

All Neurobagel services are [containerized](https://www.docker.com/resources/what-container/).
The production deployment recipes use [Docker Compose profiles](https://docs.docker.com/compose/how-tos/profiles/)
to group related services and allow them to be launched together.

### Services

The Neurobagel Docker Compose recipe includes several services

> and coordinates them to work together:

Reviewer: ```suggestion
Neurobagel services that are included in the Docker Compose deployment recipes:
```
To reduce repetition with your first sentence above, and also distinguish this list from the third-party services you list below

    Neurobagel nodes to provide a single point of access to multiple distributed databases.
    By default it will federate over all public nodes and any local nodes you specify.
- **[Neurobagel query tool](query_tool.md)** (`query_federation`): A web app that provides a graphical interface for users to query a
    federation API and view the results from one or more nodes. Because the query tool is a static app and is run locally
    in the user's browser, this service simply hosts the app.

Two additional, third-party services are part of production deployment recipes:

- **[NGINX reverse proxy](https://nginx.org/en/)** (`nginx-proxy`):
    A [containerized version](https://github.com/nginx-proxy/nginx-proxy)
    of the popular proxy server that lets you serve your Neurobagel services
    under custom URLs, handling routes automatically.
- **Automatic SSL certificate service** (`acme-companion`):
    A [containerized companion tool](https://github.com/nginx-proxy/acme-companion)
    for nginx that automatically provisions SSL certificates for the routes
    created by `nginx` so users can communicate with your services
    via encrypted HTTPS connections.

> ## Production deployment

Reviewer: ```suggestion
## Setting up a production deployment
```
so readers can easily find where the setup instructions begin

> ### Deployment profiles

Reviewer: I'd lift this section out to the same heading level as "Deployable services", since it reads to me like a reference rather than something a user needs to read every time. 

That way, this section ("Setting up a production deployment") becomes focused on just instructions (common setup and profile-specific).

Neurobagel offers different deployment profiles that allow you to launch
specific combinations of services (listed below), depending on your use case.

> - [`proxy`](#proxy-server): Deploys pre-configured, containerized

Reviewer: 🍒 I understand this is listed first to match the order of deployment, but in the spirit of this being a reference section, we could consider moving after it `node` and `portal` so the Neurobagel service profiles appear first

    to your Neurobagel services under your desired URLs.

    ??? info "You can also use an existing proxy server"

        Our deployment instructions assume that there is no existing proxy server set up
        on the machine that will host your Neurobagel services.
        If you already have a proxy server setup, 
        follow the slightly modified steps described in [deploying with an existing proxy server](#deploying-with-an-existing-proxy-server).
        
        In this case, you can ignore the `proxy` deployment recipe.
- [`node`](#node): Deploys an individual Neurobagel node.
A Neurobagel node includes an internal graph database and a node API
that handles all incoming queries and talks to the graph database.
You can run several nodes on the same machine.
       - `api`
       - `graph`

- [`portal`](#portal): Deploys the federation engine and a connected web query interface.
    Use this profile only if you need to host your own federated query tool,
    e.g. to federate over nodes that are not in the list of public Neurobagel nodes.
       - `federation`
       - `query_federation`

### Common setup for all deployment profiles

> !!! info "Do these steps first for each deployment profile you set up"

Reviewer: ```suggestion
!!! warning "Do these steps first for each deployment profile you set up"
```

    Each production deployment profile requires a fresh deployment recipe and begins with the same initial steps.
    Complete these steps before following the profile-specific setup instructions.

#### Clone the recipe repository

Make a fresh clone of the `recipes` repository in a location of your choice.

```bash

> git clone https://github.com/neurobagel/recipes.git my-new-deployment

Reviewer: ```suggestion
git clone https://github.com/neurobagel/recipes.git recipes
```
IIRC we decided to keep the default directory name the same (but just specified explicitly) so that users can easily identify where the deployment repo is relative to other dataset directories they may have.

```

Change `my-new-deployment` to a directory name you will recognize in the future.
Then navigate into this directory for the remaining steps.

```bash
cd my-new-deployment
```

#### Copy the template files

The recipe repository comes with templates of the `.env` and
`local_nb_nodes.json` files. Copy and rename these templates, but do not edit
the templates themselves.

```bash
cp template.env .env
cp local_nb_nodes.template.json local_nb_nodes.json
```

### Proxy server

??? warning "Skip if you already have a reverse proxy server"

    If you already have a reverse proxy server set up and want to continue using
    it, **do not follow this section** and instead continue with our guide
    on production deployment with existing proxy servers.

    # TODO: link

??? warning "Always launch the proxy server first"

    Our reverse proxy recipe is set up to automatically configure routes to the
    Neurobagel services you launch. In order to do that, the proxy server must
    already be running when you launch a new Neurobagel service.
    If you have already launched Neurobagel services
    (e.g. [node](#node) or [portal](#portal)),

>     shut them down again, launch the proxy server, and then relaunch the services.

Reviewer: ```suggestion
    If you have accidentally already launched production Neurobagel services
    (e.g. [node](#node) or [portal](#portal)),
    first shut them down using `docker compose down`, launch the proxy server, and then relaunch the services.
```
to differentiate from the case where they have only set up a test deployment, which should not affect the production setup steps.

!!! info "Start from a [fresh deployment recipe](#common-setup-for-all-deployment-profiles)!"

To host your Neurobagel [node services](#services) under a custom URL
(e.g. `https://www.myfirstnode.org/query`) rather than a server IP address and port

> (e.g. `http://192.168.0.1:3000`), we provide a recipe for you to easily set up an

Reviewer: ```suggestion
To host your Neurobagel [node services](#services) under custom URLs
(e.g. `https://www.myfirstnode.org/query`) rather than server IP addresses and ports
(e.g. `http://192.168.0.1:3000`), we provide a recipe for you to easily set up an
```

[NGINX](https://nginx.org/en/docs/beginners_guide.html) reverse proxy
alongside your Neurobagel services.

Make sure that:

- you have access to the domain you want to host your services at,

> so that you can create a DNS entry that points at the webserver

Reviewer: ```suggestion
so that you can create a DNS entry that points at the machine
```
we use "machine" elsewhere to describe the host server for Neurobagel services, so I'd keep the language consistent (i.e., we don't mention "webserver" anywhere else)

that will host your Neurobagel services.
- your web server firewall allows incoming connections
on ports 80 (HTTP) and 443 (HTTPS).
[Both are necessary](https://letsencrypt.org/docs/integration-guide/#firewall-configuration)
to complete the SSL certificate challenge and offer HTTPS connections.

Launch the proxy server using the corresponding deployment recipe:

```bash
docker compose -f docker-compose.proxy.yml up -d
```

You should now see the `nginx` and `nginx-acme` services running:

```bash
docker ps
```

### Node

!!! note "Start from a [fresh deployment recipe](#common-setup-for-all-deployment-profiles)!"

> 

Reviewer: This page is missing a step about adding your own data to the graph. I think we can keep it simple, something like
```suggestion
#### Add data to the node

By default, any JSONLD data in the [`./data`](https://github.com/neurobagel/recipes/tree/main/data) subdirectory of your deployment recipe directory will be automatically uploaded to the graph store.

To add the dataset JSONLD files for your node, you can either:
- place the JSONLD files inside `./data` (delete the example JSONLD provided by default with the recipe), OR
- change the path where the deployment recipe will look for JSONLD files by editing the variable `LOCAL_GRAPH_DATA` in your `.env` file
```

#### Set graph store credentials

!!! danger "This is a security relevant section!"

The graph store (GraphDB instance) in a Neurobagel node
is secured with password-based access and includes two users:
an `admin` superuser and a regular database user,
both of which are automatically configured by the Neurobagel deployment recipe.
Passwords for both users are defined via files in the `./secrets` directory
of the recipes repository, while the regular database username is set
through an environment variable in `.env` file.

??? warning "Changing user credentials after the first launch requires a hard reset"

    If you've previously launched a Neurobagel deployment (Docker Compose stack),
    you'll need to
    [reset your graph store](maintaining.md#resetting-your-graphdb-instance)

>     for any changes you have made to user credentials to take effect. 

Reviewer: ```suggestion
    To change user credentials for a Neurobagel `node` profile deployment you previously launched,
    you'll need to
    [reset your graph store](maintaining.md#resetting-your-graphdb-instance)
    for any changes to take effect. 
```

docker compose -f docker-compose.prod.yml up -d
```

??? info "Make sure the proxy service is already running"

    The default `node` deployment recipe requires that you have already
    [deployed the proxy server](#proxy-server).

### Portal

!!! note "Start from a [fresh deployment recipe](#common-setup-for-all-deployment-profiles)!"

To host your own query portal that federates over a set of nodes,
use your `local_nb_nodes.json` to configure the nodes of interest.

Each node to be federated over is defined using a dictionary with two required keys:

- `NodeName`: Display name of the node, which will be shown in the query portal
- `ApiURL`: URL of the **node API** for the node

`local_nb_nodes.json`:

    [
        {
            "NodeName": "Parkinson's Disease Data - Site 1",
            "ApiURL": "https://mydomain.org/site1"
        }
        {
            "NodeName": "Parkinson's Disease Data - Site 2",
            "ApiURL": "https://mydomain.org/site2"
        }
    ]

!!! warning "`ApiURL` must include the protocol (`http://` or `https://`)"

??? info "Public Neurobagel nodes do not need to be included"

        We maintain a list of publicly accessible Neurobagel nodes 

>     [here](https://github.com/neurobagel/menu/blob/main/node_directory/neurobagel_public_nodes.json).

Reviewer: ```suggestion
    We maintain a list of publicly accessible Neurobagel nodes 
    [here](https://github.com/neurobagel/menu/blob/main/node_directory/neurobagel_public_nodes.json).
```
weird snippet when rendered
